iroh-unixfs/src/builder.rs

```diff
@@ -520,21 +557,131 @@ impl DirectoryBuilder {
+///
+/// Either a branch or a leaf. Root will always be a branch,
+/// even if it has only one child.
+#[derive(Debug, PartialEq)]
+enum HamtNode {
+    Branch(BTreeMap<u32, HamtNode>),
+    Leaf(HamtLeaf),
+}
+
+impl HamtNode {
+    fn new(entries: Vec<Entry>) -> anyhow::Result<HamtNode> {
+        // add the hash
+        let entries = entries
+            .into_iter()
+            .map(|entry| {
+                let name = entry.name().to_string();
+                let hash = hash_key(name.as_bytes());
+                HamtLeaf(hash, entry)
+            })
+            .collect::<Vec<_>>();
+        Self::group(entries, 0, 8)
+    }
+
+    fn group(leafs: Vec<HamtLeaf>, pos: u32, len: u32) -> anyhow::Result<HamtNode> {
+        Ok(if leafs.len() == 1 && pos > 0 {
+            HamtNode::Leaf(leafs.into_iter().next().unwrap())
+        } else {
+            let mut res = BTreeMap::<u32, Vec<HamtLeaf>>::new();
+            for leaf in leafs {
+                let value = bits(&leaf.0, pos, len)?;
+                res.entry(value).or_default().push(leaf);
+            }
+            let res = res
+                .into_iter()
+                .map(|(key, leafs)| {
+                    let node = Self::group(leafs, pos + len, len)?;
+                    anyhow::Ok((key, node))
+                })
+                .collect::<anyhow::Result<_>>()?;
+            HamtNode::Branch(res)
+        })
+    }
+
+    fn name(&self) -> &str {
+        match self {
+            HamtNode::Branch(_) => "",
+            HamtNode::Leaf(HamtLeaf(_, entry)) => entry.name(),
+        }
+    }
+
+    pub fn encode<'a>(self) -> LocalBoxStream<'a, Result<Block>> {
+        match self {
+            Self::Branch(tree) => {
+                async_stream::try_stream! {
+                    let mut links = Vec::with_capacity(tree.len());
+                    let mut bitfield = Bitfield::default();
+                    for (prefix, node) in tree {
+                        let name = format!("{:02X}{}", prefix, node.name());
+                        bitfield.set_bit(prefix);
+                        let blocks = node.encode();
+                        let mut root = None;
+                        tokio::pin!(blocks);
+                        while let Some(block) = blocks.next().await {
+                            let block = block?;
+                            root = Some(*block.cid());
+                            yield block;
+                        }
+                        links.push(crate::unixfs::dag_pb::PbLink {
+                            name: Some(name),
+                            hash: root.map(|cid| cid.to_bytes()),
+                            tsize: None,
+                        });
+                    }
+                    let inner = unixfs_pb::Data {
+                        r#type: DataType::HamtShard as i32,
+                        hash_type: Some(HamtHashFunction::Murmur3 as u64),
+                        fanout: Some(256),
```

Comment:
can you reuse the constant from hamt.rs?

Reply:
will do next pr